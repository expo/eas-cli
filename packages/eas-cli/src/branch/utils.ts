import { Client } from '../vcs/vcs';

export async function getDefaultBranchNameAsync(vcsClient: Client): Promise<string> {
  return (
    (await vcsClient.getBranchNameAsync()) || `branch-${Math.random().toString(36).substring(2, 4)}`
  );
}

export class BranchNotFoundError extends Error {
  constructor(message?: string) {
    super(message ?? 'Branch not found.');
  }
}
