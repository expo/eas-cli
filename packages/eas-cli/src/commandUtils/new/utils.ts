import path from 'path';

/**
 * Formats a directory path for display.
 * If the directory is within the current working directory, returns a relative path.
 * Otherwise, returns the absolute path.
 */
export function printDirectory(directory: string): string {
  const cwd = process.cwd();
  const absoluteDir = path.isAbsolute(directory) ? directory : path.resolve(cwd, directory);

  const relativePath = path.relative(cwd, absoluteDir);

  // If the relative path doesn't start with '..' it means it's within or at the cwd
  if (!relativePath.startsWith('..') && !path.isAbsolute(relativePath)) {
    return relativePath !== '' ? `./${relativePath}` : '.';
  }

  return absoluteDir;
}
