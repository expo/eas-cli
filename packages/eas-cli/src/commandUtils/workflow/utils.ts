import { fetchRawLogsForJobAsync } from './fetchLogs';
import {
  WorkflowJobResult,
  WorkflowLogLine,
  WorkflowLogs,
  WorkflowRunResult,
  WorkflowTriggerType,
} from './types';
import {
  WorkflowRunFragment,
  WorkflowRunStatus,
  WorkflowRunTriggerEventType,
} from '../../graphql/generated';
import { Choice } from '../../prompts';

export function computeTriggerInfoForWorkflowRun(run: WorkflowRunFragment): {
  triggerType: WorkflowTriggerType;
  trigger: string | null;
} {
  let triggerType = WorkflowTriggerType.OTHER;
  let trigger = '';
  if (run.actor?.__typename === 'Robot') {
    if (run.actor.firstName?.startsWith('GitHub App Â· ')) {
      trigger = `${run.requestedGitRef ?? ''}@${run.gitCommitHash?.substring(0, 12) ?? ''}`;
    }
  } else if (run.actor?.__typename === 'User') {
    trigger = run.actor.username;
  }
  switch (run.triggerEventType) {
    case WorkflowRunTriggerEventType.Manual:
      triggerType = WorkflowTriggerType.MANUAL;
      break;
    case WorkflowRunTriggerEventType.GithubPullRequestLabeled:
    case WorkflowRunTriggerEventType.GithubPullRequestOpened:
    case WorkflowRunTriggerEventType.GithubPullRequestReopened:
    case WorkflowRunTriggerEventType.GithubPullRequestSynchronize:
    case WorkflowRunTriggerEventType.GithubPush:
      triggerType = WorkflowTriggerType.GITHUB;
      break;
    case WorkflowRunTriggerEventType.Schedule:
      triggerType = WorkflowTriggerType.SCHEDULED;
      trigger = run.triggeringSchedule ?? '';
      break;
  }
  return { triggerType, trigger };
}

export function choiceFromWorkflowRun(run: WorkflowRunResult): Choice {
  const titleArray = [
    run.workflowFileName,
    run.status,
    run.startedAt,
    run.triggerType,
    run.trigger,
  ];
  return {
    title: titleArray.join(' - '),
    value: run.id,
    description: `ID: ${run.id}, Message: ${run.gitCommitMessage?.split('\n')[0] ?? ''}`,
  };
}

export function choiceFromWorkflowJob(job: WorkflowJobResult, index: number): Choice {
  return {
    title: `${job.name} - ${job.status}`,
    value: index,
    description: `ID: ${job.id}`,
  };
}

export function choicesFromWorkflowLogs(logs: WorkflowLogs): Choice[] {
  return Array.from(logs.keys()).map(step => {
    const logLines = logs.get(step);
    const stepStatus =
      logLines?.filter((line: WorkflowLogLine) => line.marker === 'end-step')[0]?.result ?? '';
    return {
      title: `${step} - ${stepStatus}`,
      value: step,
    };
  });
}

export function processWorkflowRuns(runs: WorkflowRunFragment[]): WorkflowRunResult[] {
  return runs.map(run => {
    const finishedAt = run.status === WorkflowRunStatus.InProgress ? null : run.updatedAt;
    const { triggerType, trigger } = computeTriggerInfoForWorkflowRun(run);
    return {
      id: run.id,
      status: run.status,
      gitCommitMessage: run.gitCommitMessage?.split('\n')[0] ?? null,
      gitCommitHash: run.gitCommitHash ?? null,
      startedAt: run.createdAt,
      finishedAt,
      triggerType,
      trigger,
      workflowId: run.workflow.id,
      workflowName: run.workflow.name ?? null,
      workflowFileName: run.workflow.fileName,
    };
  });
}

export async function processLogsFromJobAsync(
  job: WorkflowJobResult
): Promise<WorkflowLogs | null> {
  const rawLogs = await fetchRawLogsForJobAsync(job);
  if (!rawLogs) {
    return null;
  }
  const logs: WorkflowLogs = new Map();
  const logKeys = new Set<string>();
  rawLogs.split('\n').forEach(line => {
    try {
      const parsedLine = JSON.parse(line);
      const { result, marker } = parsedLine;
      const { buildStepDisplayName, buildStepInternalId, time, msg } = parsedLine;
      const stepId = buildStepDisplayName ?? buildStepInternalId;
      if (stepId) {
        if (!logKeys.has(stepId)) {
          logKeys.add(stepId);
          logs.set(stepId, []);
        }
        logs.get(stepId)?.push({ time, msg, result, marker });
      }
    } catch {}
  });
  return logs;
}
