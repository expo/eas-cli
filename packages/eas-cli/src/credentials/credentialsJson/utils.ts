import path from 'path';

import { Target } from '../ios/types';
import { IosCredentials } from './types';

export function getCredentialsJsonPath(projectDir: string): string {
  return path.join(projectDir, 'credentials.json');
}

export function ensureAllTargetsAreConfigured(
  targets: Target[],
  credentialsJson: IosCredentials
): void {
  const notConfiguredTargets: string[] = [];
  for (const target of targets) {
    if (!(target.targetName in credentialsJson)) {
      notConfiguredTargets.push(target.targetName);
      continue;
    }
  }

  if (notConfiguredTargets.length > 0) {
    const errorMessage =
      targets.length === 1
        ? 'Credentials are not defined in credentials.json'
        : `Credentials for target${
            notConfiguredTargets.length === 1 ? '' : 's'
          } ${notConfiguredTargets
            .map(i => `'${i}'`)
            .join(',')} are not defined in credentials.json`;
    throw new Error(errorMessage);
  }
}
