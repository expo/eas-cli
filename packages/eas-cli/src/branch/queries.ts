import chalk from 'chalk';
import CliTable from 'cli-table3';

import { PaginatedQueryOptions } from '../commandUtils/pagination';
import { UpdateBranchFragment } from '../graphql/generated';
import { BranchQuery } from '../graphql/queries/BranchQuery';
import Log from '../log';
import { UPDATE_COLUMNS, formatUpdateMessage, getPlatformsForGroup } from '../update/utils';
import { printJsonOnlyOutput } from '../utils/json';
import {
  paginatedQueryWithConfirmPromptAsync,
  paginatedQueryWithSelectPromptAsync,
} from '../utils/queries';

export const BRANCHES_LIMIT = 50;

export async function selectBranchOnAppAsync({
  projectId,
  promptTitle,
  displayTextForListItem,
  paginatedQueryOptions,
}: {
  projectId: string;
  displayTextForListItem: (queryItem: UpdateBranchFragment) => string;
  promptTitle: string;
  paginatedQueryOptions: PaginatedQueryOptions;
}): Promise<UpdateBranchFragment> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select a branch in non-interactive mode.');
  }

  const selectedBranch = await paginatedQueryWithSelectPromptAsync({
    limit: paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
    offset: paginatedQueryOptions.offset,
    queryToPerform: (limit, offset) => queryBranchesOnProjectAsync(limit, offset, projectId),
    promptOptions: {
      title: promptTitle,
      getIdentifierForQueryItem: updateBranchFragment => updateBranchFragment.id,
      createDisplayTextForSelectionPromptListItem: displayTextForListItem,
    },
  });
  if (!selectedBranch) {
    throw new Error(`Could not find any branches for project "${projectId}"`);
  }
  return selectedBranch;
}

export async function listAndRenderBranchesOnAppAsync({
  projectId,
  paginatedQueryOptions,
}: {
  projectId: string;
  paginatedQueryOptions: PaginatedQueryOptions;
}): Promise<void> {
  if (paginatedQueryOptions.nonInteractive) {
    const branches = await queryBranchesOnProjectAsync(
      paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
      paginatedQueryOptions.offset,
      projectId
    );
    renderPageOfBranches(branches, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? BRANCHES_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) => queryBranchesOnProjectAsync(limit, offset, projectId),
      promptOptions: {
        title: 'Load more branches?',
        renderListItems: branches => renderPageOfBranches(branches, paginatedQueryOptions),
      },
    });
  }
}

async function queryBranchesOnProjectAsync(
  limit: number,
  offset: number,
  projectId: string
): Promise<UpdateBranchFragment[]> {
  return await BranchQuery.listBranchesOnAppAsync({
    appId: projectId,
    limit,
    offset,
  });
}

function renderPageOfBranches(
  currentPage: UpdateBranchFragment[],
  { json }: PaginatedQueryOptions
): void {
  if (json) {
    printJsonOnlyOutput(currentPage);
  } else {
    const table = new CliTable({
      head: ['Branch', ...UPDATE_COLUMNS],
      wordWrap: true,
    });

    table.push(
      ...currentPage.map(branch => {
        if (branch.updates.length === 0) {
          return [branch.name, 'N/A', 'N/A', 'N/A', 'N/A'];
        }

        const latestUpdateOnBranch = branch.updates[0];
        return [
          branch.name,
          formatUpdateMessage(latestUpdateOnBranch),
          latestUpdateOnBranch.runtimeVersion,
          latestUpdateOnBranch.group,
          getPlatformsForGroup({
            group: latestUpdateOnBranch.group,
            updates: branch.updates,
          }),
        ];
      })
    );

    Log.addNewLineIfNone();
    Log.log(chalk.bold('Branches:'));
    Log.addNewLineIfNone();
    Log.log(table.toString());
  }
}
