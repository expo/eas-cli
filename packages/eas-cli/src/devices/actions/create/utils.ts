import chalk from 'chalk';

import { AppleDeviceClass, AppleTeam } from '../../../graphql/generated';
import Log from '../../../log';
import { promptAsync } from '../../../prompts';
import { isValidUDID, normalizeUDID } from '../../udids';
import { formatNewDevice } from '../../utils/formatDevice';

export interface DeviceData {
  udid: string;
  name?: string;
  deviceClass: AppleDeviceClass | null;
}

export async function promptForDeviceClassAsync(
  initial?: AppleDeviceClass | null
): Promise<AppleDeviceClass | null> {
  const choices = [
    { title: 'iPhone', value: AppleDeviceClass.Iphone },
    { title: 'iPad', value: AppleDeviceClass.Ipad },
    { title: 'Mac', value: AppleDeviceClass.Mac },
    { title: 'Not sure (leave empty)', value: null },
  ];
  const values = choices.map(({ value }) => value);

  const { deviceClass } = await promptAsync({
    type: 'select',
    name: 'deviceClass',
    message: 'Device class (optional):',
    choices,
    initial: initial !== undefined && values.indexOf(initial),
  });
  return deviceClass;
}

export async function promptForNameAsync(initial?: string): Promise<string | undefined> {
  const { name } = await promptAsync({
    type: 'text',
    name: 'name',
    message: 'Device name (optional):',
    initial,
  });
  return name;
}

export async function promptForUDIDAsync(initial?: string): Promise<string> {
  const { udid } = await promptAsync({
    type: 'text',
    name: 'udid',
    message: 'UDID:',
    initial,
    validate: (rawVal: string) => {
      const val = normalizeUDID(rawVal);
      if (!val || val === '') {
        return 'UDID cannot be empty';
      } else if (val.length !== 25 && val.length !== 40) {
        return 'UDID should be a 25 or 40-character string';
      } else if (!isValidUDID(val)) {
        return 'UDID is invalid';
      } else {
        return true;
      }
    },
    format: (val: string) => normalizeUDID(val),
  });
  return udid;
}

export function printDeviceData(
  deviceData: DeviceData,
  appleTeam: Pick<AppleTeam, 'appleTeamIdentifier' | 'appleTeamName'>
): void {
  Log.newLine();
  Log.log(
    `We are going to register the following device in our database.
  This device will ${chalk.bold(
    'not'
  )} be registered on the Apple Developer Portal until it is chosen for an internal distribution build.`
  );
  Log.newLine();
  Log.log(formatNewDevice({ ...deviceData, identifier: deviceData.udid }, appleTeam));
  Log.newLine();
}
