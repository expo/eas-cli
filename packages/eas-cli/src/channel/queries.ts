import chalk from 'chalk';
import gql from 'graphql-tag';

import { ExpoGraphqlClient } from '../commandUtils/context/contextUtils/createGraphqlClient';
import { PaginatedQueryOptions } from '../commandUtils/pagination';
import { withErrorHandlingAsync } from '../graphql/client';
import {
  CreateUpdateChannelOnAppMutation,
  CreateUpdateChannelOnAppMutationVariables,
  ViewBranchesOnUpdateChannelQueryVariables,
  ViewUpdateChannelsOnAppQueryVariables,
} from '../graphql/generated';
import { BranchQuery, UpdateBranchOnChannelObject } from '../graphql/queries/BranchQuery';
import { ChannelQuery, UpdateChannelObject } from '../graphql/queries/ChannelQuery';
import Log from '../log';
import formatFields from '../utils/formatFields';
import { printJsonOnlyOutput } from '../utils/json';
import {
  paginatedQueryWithConfirmPromptAsync,
  paginatedQueryWithSelectPromptAsync,
} from '../utils/queries';
import { logChannelDetails } from './utils';

export const CHANNELS_LIMIT = 25;

export async function selectChannelOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId,
    selectionPromptTitle,
    paginatedQueryOptions,
  }: {
    projectId: string;
    selectionPromptTitle: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<UpdateChannelObject> {
  if (paginatedQueryOptions.nonInteractive) {
    throw new Error('Unable to select a channel in non-interactive mode.');
  }

  const updateChannel = await paginatedQueryWithSelectPromptAsync({
    limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
    offset: paginatedQueryOptions.offset,
    queryToPerform: (limit, offset) =>
      queryChannelsOnAppAsync(graphqlClient, { appId: projectId, limit, offset }),
    promptOptions: {
      title: selectionPromptTitle,
      makePartialChoiceObject: updateChannel => ({ title: updateChannel.name }),
      getIdentifierForQueryItem: updateChannel => updateChannel.id,
    },
  });

  if (!updateChannel) {
    throw new Error(`Could not find any channels for app "${projectId}"`);
  }
  return updateChannel;
}

export async function listAndRenderChannelsOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId,
    paginatedQueryOptions,
  }: {
    projectId: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<void> {
  if (paginatedQueryOptions.nonInteractive) {
    const channels = await queryChannelsOnAppAsync(graphqlClient, {
      appId: projectId,
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
    });
    renderPageOfChannels(channels, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        queryChannelsOnAppAsync(graphqlClient, { limit, offset, appId: projectId }),
      promptOptions: {
        title: 'Load more channels?',
        renderListItems: channels => renderPageOfChannels(channels, paginatedQueryOptions),
      },
    });
  }
}

export async function listAndRenderBranchesAndUpdatesOnChannelAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    projectId: appId,
    channelName,
    paginatedQueryOptions,
  }: {
    projectId: string;
    channelName: string;
    paginatedQueryOptions: PaginatedQueryOptions;
  }
): Promise<void> {
  const channel = await ChannelQuery.viewUpdateChannelAsync(graphqlClient, { appId, channelName });
  renderChannelHeaderContent({ channelName: channel.name, channelId: channel.id });

  if (paginatedQueryOptions.nonInteractive) {
    const branches = await queryBranchesAndUpdateGroupsOnChannelAsync(graphqlClient, {
      appId,
      channelName,
      offset: paginatedQueryOptions.offset,
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
    });
    renderPageOfBranchesOnChannel(channel, branches, paginatedQueryOptions);
  } else {
    await paginatedQueryWithConfirmPromptAsync({
      limit: paginatedQueryOptions.limit ?? CHANNELS_LIMIT,
      offset: paginatedQueryOptions.offset,
      queryToPerform: (limit, offset) =>
        queryBranchesAndUpdateGroupsOnChannelAsync(graphqlClient, {
          channelName,
          appId,
          offset,
          limit,
        }),
      promptOptions: {
        title: 'Load more channels?',
        renderListItems: branches =>
          renderPageOfBranchesOnChannel(channel, branches, paginatedQueryOptions),
      },
    });
  }
}

async function queryChannelsOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  { appId, offset, limit }: ViewUpdateChannelsOnAppQueryVariables
): Promise<UpdateChannelObject[]> {
  return await ChannelQuery.viewUpdateChannelsOnAppAsync(graphqlClient, {
    appId,
    offset,
    limit,
  });
}

async function queryBranchesAndUpdateGroupsOnChannelAsync(
  graphqlClient: ExpoGraphqlClient,
  args: ViewBranchesOnUpdateChannelQueryVariables
): Promise<UpdateBranchOnChannelObject[]> {
  return await BranchQuery.listBranchesOnChannelAsync(graphqlClient, args);
}

function renderPageOfChannels(
  currentPage: UpdateChannelObject[],
  { json }: PaginatedQueryOptions
): void {
  if (json) {
    printJsonOnlyOutput({ currentPage });
  } else {
    for (const channel of currentPage) {
      renderChannelHeaderContent({ channelName: channel.name, channelId: channel.id });
      Log.addNewLineIfNone();
      logChannelDetails(channel);

      if (currentPage.indexOf(channel) < currentPage.length - 1) {
        Log.log(`\n${chalk.dim('———')}\n`);
      }
    }
  }
}

function renderPageOfBranchesOnChannel(
  channel: UpdateChannelObject,
  currentPage: UpdateBranchOnChannelObject[],
  { json }: PaginatedQueryOptions
): void {
  const channelWithNewBranches = { ...channel, updateBranches: currentPage };
  if (json) {
    printJsonOnlyOutput({ currentPage: channelWithNewBranches });
  } else {
    // The channel details contain both the branch and latest update group
    Log.addNewLineIfNone();
    logChannelDetails(channelWithNewBranches);
  }
}

function renderChannelHeaderContent({
  channelName,
  channelId,
}: {
  channelName: string;
  channelId: string;
}): void {
  Log.addNewLineIfNone();
  Log.log(chalk.bold('Channel:'));
  Log.log(
    formatFields([
      { label: 'Name', value: channelName },
      { label: 'ID', value: channelId },
    ])
  );
  Log.addNewLineIfNone();
  Log.log(chalk`{bold Branches pointed at this channel and their most recent update group:}`);
}

export async function createChannelOnAppAsync(
  graphqlClient: ExpoGraphqlClient,
  {
    appId,
    branchId,
    channelName,
  }: {
    appId: string;
    branchId: string;
    channelName: string;
  }
): Promise<CreateUpdateChannelOnAppMutation> {
  // Point the new channel at a branch with its same name.
  const branchMapping = JSON.stringify({
    data: [{ branchId, branchMappingLogic: 'true' }],
    version: 0,
  });
  return await withErrorHandlingAsync(
    graphqlClient
      .mutation<CreateUpdateChannelOnAppMutation, CreateUpdateChannelOnAppMutationVariables>(
        gql`
          mutation CreateUpdateChannelOnApp($appId: ID!, $name: String!, $branchMapping: String!) {
            updateChannel {
              createUpdateChannelForApp(appId: $appId, name: $name, branchMapping: $branchMapping) {
                id
                name
                branchMapping
              }
            }
          }
        `,
        {
          appId,
          name: channelName,
          branchMapping,
        }
      )
      .toPromise()
  );
}

export async function ensureChannelExistsAsync(
  graphqlClient: ExpoGraphqlClient,
  { appId, branchId, channelName }: { appId: string; branchId: string; channelName: string }
): Promise<void> {
  try {
    await createChannelOnAppAsync(graphqlClient, {
      appId,
      channelName,
      branchId,
    });
  } catch (e: any) {
    const isIgnorableError =
      e.graphQLErrors?.length === 1 &&
      e.graphQLErrors[0].extensions.errorCode === 'CHANNEL_ALREADY_EXISTS';
    if (!isIgnorableError) {
      throw e;
    }
  }
}
