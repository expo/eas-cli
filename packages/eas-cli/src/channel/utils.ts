import assert from 'assert';
import chalk from 'chalk';

import { UpdateChannelObject } from '../graphql/queries/ChannelQuery';
import Log from '../log';
import {
  FormattedBranchDescription,
  formatBranch,
  getUpdateGroupDescriptionsWithBranch,
} from '../update/utils';

export type BranchMapping = {
  version: number;
  data: {
    branchId: string;
    branchMappingLogic: {
      operand: number;
      clientKey: string;
      branchMappingOperator: string;
    } & string;
  }[];
};

/**
 * Get the branch mapping and determine whether it is a rollout.
 * Ensure that the branch mapping is properly formatted.
 */
export function getBranchMapping(branchMappingString?: string): {
  branchMapping: BranchMapping;
  isRollout: boolean;
  rolloutPercent?: number;
} {
  if (!branchMappingString) {
    throw new Error('Missing branch mapping.');
  }
  let branchMapping: BranchMapping;
  try {
    branchMapping = JSON.parse(branchMappingString);
  } catch {
    throw new Error(`Could not parse branchMapping string into a JSON: "${branchMappingString}"`);
  }
  assert(branchMapping, 'Branch Mapping must be defined.');

  if (branchMapping.version !== 0) {
    throw new Error('Branch mapping must be version 0.');
  }

  const isRollout = branchMapping.data.length === 2;
  const rolloutPercent = branchMapping.data[0]?.branchMappingLogic.operand;

  switch (branchMapping.data.length) {
    case 0:
      break;
    case 1:
      if (branchMapping.data[0].branchMappingLogic !== 'true') {
        throw new Error('Branch mapping logic for a single branch must be "true"');
      }
      break;
    case 2:
      if (branchMapping.data[0].branchMappingLogic.clientKey !== 'rolloutToken') {
        throw new Error('Client key of initial branch mapping must be "rolloutToken"');
      }
      if (branchMapping.data[0].branchMappingLogic.branchMappingOperator !== 'hash_lt') {
        throw new Error('Branch mapping operator of initial branch mapping must be "hash_lt"');
      }
      if (rolloutPercent == null) {
        throw new Error('Branch mapping is missing a "rolloutPercent"');
      }
      if (branchMapping.data[1].branchMappingLogic !== 'true') {
        throw new Error('Branch mapping logic for a the second branch of a rollout must be "true"');
      }
      break;
    default:
      throw new Error('Branch mapping data must have length less than or equal to 2.');
  }

  return { branchMapping, isRollout, rolloutPercent };
}

export function logChannelDetails(channel: UpdateChannelObject): void {
  const { branchMapping, isRollout, rolloutPercent } = getBranchMapping(channel.branchMapping);
  if (branchMapping.data.length > 2) {
    throw new Error('Branch Mapping data must have length less than or equal to 2.');
  }

  const rolloutBranchIds = branchMapping.data.map(data => data.branchId);
  const branchDescription = channel.updateBranches.flatMap(branch => {
    const updateGroupWithBranchDescriptions = getUpdateGroupDescriptionsWithBranch(
      branch.updateGroups
    );

    const isRolloutBranch = isRollout && rolloutBranchIds.includes(branch.id);
    const isBaseBranch = rolloutBranchIds.length > 0 && rolloutBranchIds[0] === branch.id;
    let rolloutPercentNumber: number | undefined = undefined;
    if (isRolloutBranch) {
      rolloutPercentNumber = isBaseBranch ? rolloutPercent! * 100 : (1 - rolloutPercent!) * 100;
    }

    return updateGroupWithBranchDescriptions.map(
      ({ branch, ...updateGroup }): FormattedBranchDescription => ({
        branch,
        branchRolloutPercentage: rolloutPercentNumber,
        update: updateGroup,
      })
    );
  });

  if (branchDescription.length === 0) {
    Log.log(chalk.dim('No branches are pointed to this channel.'));
  } else {
    Log.log(
      branchDescription
        .map(description => formatBranch(description))
        .join(`\n\n${chalk.dim('———')}\n\n`)
    );
  }
}
