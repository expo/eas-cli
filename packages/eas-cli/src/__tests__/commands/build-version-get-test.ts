import { AppVersionSource, EasJson } from '@expo/eas-json';
import chalk from 'chalk';

import { getMockEasJson, mockCommandContext, mockProjectId, mockTestCommand } from './utils';
import BuildVersionGetView from '../../commands/build/version/get';
import { AppVersionQuery } from '../../graphql/queries/AppVersionQuery';
import Log from '../../log';
import { enableJsonOutput, printJsonOnlyOutput } from '../../utils/json';

jest.mock('../../project/applicationIdentifier');
jest.mock('fs');
jest.mock('../../log');
jest.mock('../../utils/json');
jest.mock('../../graphql/queries/AppVersionQuery');

function withRemoteVersionSource(easJson: EasJson): EasJson {
  return {
    ...easJson,
    cli: {
      ...easJson.cli,
      appVersionSource: AppVersionSource.REMOTE,
    },
  };
}

describe(BuildVersionGetView, () => {
  afterEach(() => {
    jest.clearAllMocks();
  });
  test('reading version for platform android', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {
      easJson: withRemoteVersionSource(getMockEasJson()),
    });
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => ({
      buildVersion: '100',
      storeVersion: '1.0.0',
    }));

    const cmd = mockTestCommand(BuildVersionGetView, ['--platform=android'], ctx);
    await cmd.run();
    expect(AppVersionQuery.latestVersionAsync).toHaveBeenCalledWith(
      ctx.loggedIn.graphqlClient,
      mockProjectId,
      'ANDROID',
      'eas.test.com'
    );
    expect(Log.log).toHaveBeenCalledWith(`Android versionCode - ${chalk.bold('100')}`);
    expect(enableJsonOutput).not.toHaveBeenCalled();
    expect(printJsonOnlyOutput).not.toHaveBeenCalled();
  });

  test('reading version for platform android when no remote version is set', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {
      easJson: withRemoteVersionSource(getMockEasJson()),
    });
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => null);

    const cmd = mockTestCommand(BuildVersionGetView, ['--platform=android'], ctx);
    await cmd.run();
    expect(AppVersionQuery.latestVersionAsync).toHaveBeenCalledWith(
      ctx.loggedIn.graphqlClient,
      mockProjectId,
      'ANDROID',
      'eas.test.com'
    );
    expect(Log.log).toHaveBeenCalledWith(`No remote versions are configured for this project.`);
    expect(enableJsonOutput).not.toHaveBeenCalled();
    expect(printJsonOnlyOutput).not.toHaveBeenCalled();
  });

  test('reading version with --json flag for platform android', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {
      easJson: withRemoteVersionSource(getMockEasJson()),
    });
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => ({
      buildVersion: '100',
      storeVersion: '1.0.0',
    }));

    const cmd = mockTestCommand(
      BuildVersionGetView,
      ['--non-interactive', '--json', '--platform=android'],
      ctx
    );
    await cmd.run();
    expect(AppVersionQuery.latestVersionAsync).toHaveBeenCalledWith(
      ctx.loggedIn.graphqlClient,
      mockProjectId,
      'ANDROID',
      'eas.test.com'
    );
    expect(enableJsonOutput).toHaveBeenCalled();
    expect(printJsonOnlyOutput).toHaveBeenCalledWith({
      versionCode: '100',
    });
  });

  test('reading version with --json flag for platform ios', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {
      easJson: withRemoteVersionSource(getMockEasJson()),
    });
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => ({
      buildVersion: '100',
      storeVersion: '1.0.0',
    }));

    const cmd = mockTestCommand(
      BuildVersionGetView,
      ['--non-interactive', '--json', '--platform=ios'],
      ctx
    );
    await cmd.run();
    expect(AppVersionQuery.latestVersionAsync).toHaveBeenCalledWith(
      ctx.loggedIn.graphqlClient,
      mockProjectId,
      'IOS',
      'eas.test.com'
    );
    expect(enableJsonOutput).toHaveBeenCalled();
    expect(printJsonOnlyOutput).toHaveBeenCalledWith({
      buildNumber: '100',
    });
  });

  test('reading version with --json flag when no remote version is set', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {
      easJson: withRemoteVersionSource(getMockEasJson()),
    });
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => null);

    const cmd = mockTestCommand(
      BuildVersionGetView,
      ['--non-interactive', '--json', '--platform=android'],
      ctx
    );
    await cmd.run();
    expect(AppVersionQuery.latestVersionAsync).toHaveBeenCalledWith(
      ctx.loggedIn.graphqlClient,
      mockProjectId,
      'ANDROID',
      'eas.test.com'
    );
    expect(enableJsonOutput).toHaveBeenCalled();
    expect(printJsonOnlyOutput).toHaveBeenCalledWith({});
  });

  test('reading version when appVersionSource is set to local ', async () => {
    const ctx = mockCommandContext(BuildVersionGetView, {});
    jest.mocked(AppVersionQuery.latestVersionAsync).mockImplementation(async () => ({
      buildVersion: '100',
      storeVersion: '1.0.0',
    }));

    const cmd = mockTestCommand(
      BuildVersionGetView,
      ['--non-interactive', '--json', '--platform=android'],
      ctx
    );
    await expect(cmd.run()).rejects.toThrowErrorMatchingSnapshot();
  });
});
