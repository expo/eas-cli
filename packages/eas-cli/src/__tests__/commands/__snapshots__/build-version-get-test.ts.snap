// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildVersionGetView reading version aborts when the appVersionSource is not specified and the user chooses to set it to LOCAL 1`] = `"Aborting..."`;

exports[`BuildVersionGetView reading version when appVersionSource is set to local  1`] = `"This project is not configured for using remote version source. Add [1m{"cli": { "appVersionSource": "remote" }}[22m in eas.json or re-run this command without "--non-interactive" flag."`;
